### ISSUE TITLE
Support Node 18.x

### ISSUE THREAD

**Issue by air2** [aws/elastic-beanstalk-roadmap#241]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave "+1" or "me too" comments. They generate extra noise for issue followers and do not help prioritize the request.

**Tell us about your request**
Add support for node 18.x. We want to move to the next node version ASAP, because it brings some new features we want to use, but we are also depended on elastic beanstalk to run our software.

**Comment by timewarrior** [comment]
Hi, Do we have a date on these. We are looking forward to move to the Node v18!

**Comment by pniaz** [comment]
Hi, Do we have a date on these. We are looking forward to move to the Node v18!

**Comment by timewarrior** [comment]
Instead of waiting we moved to Fargate. It was very easy. Took about 1-2 days to familiarize and then less amount of Pulumi script. If you don't want to use Pulumi/Terraform - you can use Copilot.
Overall Fargate is so much better than Elastic Beanstalk. System is so much more stable. Don't need to worry about platform upgrades. Node.js upgrades are controlled by me. Don't need to worry about elastic beanstalk internals. No nginx in picture. Uses less memory and CPU, so saving more compared to Elastic Beanstalk. Less security risks. No need to worry about unhelathy hosts during deploy messing up production. Red/Blue deploys with Fargate are much faster than Elastic Beanstalk.

**Comment by thiagojedi** [comment]
Just to add some info:

Node v18 is now the "Active LTS", replacing the previous v16. 

Source: https://nodejs.org/en/blog/release/v18.12.0/

**Comment by chukwumaokere** [comment]
Still waiting on this!

**Comment by ghost** [comment]
Any update on this? In future we're probably going to move away from beanstalk as the updates seem slow.

**Comment by chukwumaokere** [comment]
I had a talk with a rep at AWS and they said that their schedule and slowed them down on updating their APIs and platform versions due to a lot of things happening at AWS. They are committed to catching back up next year. However, they will likely not get to it until May at the earliest (but I wouldn't count on this).

Im looking towards using containers on fargate now 😓

**Comment by ShqipognjaLL** [comment]
Is there a potential workaround in the meantime with beanstalk configs and NVM?

**Comment by hayd** [comment]
> Is there a potential workaround in the meantime with beanstalk configs and NVM?

Best bet might be to use [docker](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/create_deploy_docker.html).

That said, I'm waiting (somewhat patiently) for 18.x...

**Comment by camjc** [comment]
AWS Customers have asked for support and clarity on timing for Node support a few years in a row,
and each issue raised gets many responses and no clarity on timing until 'its live now'.

v14 https://github.com/aws/elastic-beanstalk-roadmap/issues/138 (66 responses)
v16 https://github.com/aws/elastic-beanstalk-roadmap/issues/201 (125 responses)
v18 (this issue, ~93~ 149 responses so far)

I believe that an approximate roadmap from AWS on what to expect, would greatly improve confidence in the elastic beanstalk platform and enable AWS customers to plan their upgrades (instead of all leaving the Node platform for Docker-based solutions as many above have already done).

Any support from AWS to clarify what our timeline expectations should be, is greatly appreciated.
For example, a comment saying May/June will help lots of folks plan their year as we come into EB dropping v14 support in the coming months.

Thanks in advance for anything you can share, even if its only indicative.

**Comment by ryanhugh** [comment]
Any update here? We would love to use Node.js v18 on Beanstalk!

**Comment by Jplus2** [comment]
Any update about this?

**Comment by uhagemann** [comment]
Just wanted to add that Node 16 LTS will also [run out of support by the Node.js team in September 2023](https://nodejs.org/en/blog/announcements/nodejs16-eol). So, I hope Node 18 LTS will be a supported EB platform by then...

**Comment by camjc** [comment]
For the benefit of my fellow AWS EB customers in the thread,
I'd like to share that apparently AWS released Node18 support yesterday.

https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2023-04-03-linux.html

**Comment by dgagnon-simpleset** [comment]
As a reminder for users:
aws elasticbeanstalk update-environment --environment-name "$YOUR_EB_ENV" --solution-stack-name "arn:aws:elasticbeanstalk:ca-central-1::platform/Node.js 18 running on 64bit Amazon Linux 2/5.8.0"

Change your region and EB ENV appropriately in the string.

**Comment by Ottunger** [comment]
Tried using this new platform but having some issues depending on node internal modules used:

```
[ec2-user@ip-172-31-32-152 current]$ node src/index.js dev
node:internal/modules/cjs/loader:1338
  return process.dlopen(module, path.toNamespacedPath(filename));
                 ^

Error: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by /var/app/current/node_modules/canvas/build/Release/libglib-2.0.so.0)
    at Module._extensions..node (node:internal/modules/cjs/loader:1338:18)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Object.<anonymous> (/var/app/current/node_modules/canvas/lib/bindings.js:3:18)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12) {
  code: 'ERR_DLOPEN_FAILED'
}

Node.js v18.15.0
```

**Comment by Ottunger** [comment]
Also, even after your command
`aws elasticbeanstalk update-environment --environment-name "$YOUR_EB_ENV" --solution-stack-name "arn:aws:elasticbeanstalk:ca-central-1::platform/Node.js 18 running on 64bit Amazon Linux 2/5.8.0"`
sometimes the AMI used remains `aws-elasticbeanstalk-amzn-2.0.20220719.64bit_arm-eb_nodejs14_amazon_linux_2-hvm-2022-08-03T18-47` but this also seems unreproducible whether an environment will be subject to this or not.


EDIT: when an environment is affected, it seems like it will always remain as such. Reverting it to node14 and again node18 won't help.

There is also a bug in AWS console as of right now that does not allow to select arm64 instance type at environment creation, and so we're stuck having to wait here (cannot create a new environment and decomission old one).

**Comment by Ottunger** [comment]
Another odd thing, the AMI used for nodejs18 `aws-elasticbeanstalk-amzn-2.0.20230320.64bit-eb_nodejs18_amazon_linux_2-hvm-2023-03-29T16-33` still has in `libc6` support for up to `GLIBC <= 2.27`, but `node18` officially requires `GLIBC = 2.28` in public distributions, and this makes some native modules fail (`canvas`, for instance)

**Comment by davecarlson** [comment]
After upgrade, you need to go to the AMI ID in Configuration=>Capacity.

I can see for this platform it's `ami-0dc2a466f67b684cc` (assuming you are on x64, not arm)
If you need the ARM image, it's `ami-0265660192ccd86ad`

```
{
    "PlatformDescription": {
        "PlatformArn": "arn:aws:elasticbeanstalk:eu-west-2::platform/Node.js 18 running on 64bit Amazon Linux 2/5.8.0",
        "PlatformOwner": "AWSElasticBeanstalk",
        "PlatformName": "Node.js 18 running on 64bit Amazon Linux 2",
        "PlatformVersion": "5.8.0",
        "SolutionStackName": "64bit Amazon Linux 2 v5.8.0 running Node.js 18",
        "PlatformStatus": "Ready",
        "DateCreated": "2023-04-01T06:19:13.017000+00:00",
        "DateUpdated": "2023-04-03T17:42:28.453000+00:00",
        "PlatformCategory": "Node.js",
        "Description": "64bit Amazon Linux 2 running Node.js 18",
        "Maintainer": "aws-elasticbeanstalk-platforms@amazon.com",
        "OperatingSystemName": "Amazon Linux",
        "OperatingSystemVersion": "2",
        "ProgrammingLanguages": [
            {
                "Name": "JavaScript",
                "Version": ""
            }
        ],
        "Frameworks": [
            {
                "Name": "",
                "Version": ""
            }
        ],
        "CustomAmiList": [
            {
                "VirtualizationType": "pv",
                "ImageId": ""
            },
            {
                "VirtualizationType": "hvm",
                "ImageId": "ami-0dc2a466f67b684cc"
            }
        ],
        "SupportedTierList": [
            "WebServer/Standard",
            "Worker/SQS/HTTP"
        ],
        "SupportedAddonList": [
            "Log/CloudWatch",
            "Log/S3",
            "Monitoring/Healthd",
            "WorkerDaemon/SQSD"
        ],
        "PlatformLifecycleState": "Recommended",
        "PlatformBranchName": "Node.js 18 running on 64bit Amazon Linux 2",
        "PlatformBranchLifecycleState": "Supported"
    }
}
```

**Comment by chadhobson** [comment]
Trying to move a project over to v18 but running into a weird IP provisioning issue. Not having this issue with other NodeJS versions on EB. To reproduce..

`eb create YOUR-APP-NAME --vpc --sample --instance_type t2.small`

This will deploy the sample app on a single server in a VPC which gets a v4 IP fine and all is well. However, when you change the minimum instances to 2, the second server appears to be deployed fine in EC2 but it never sends data and never gets a v4 IP assigned so its unreachable and you're stuck in a loop because the deployment never succeeds.

I've been using EB since it first came out, so this isn't a newbie problem. Can anyone confirm they're seeing this as well?

**Comment by darin-holloway** [comment]
This is done

**Comment by coval3nte** [comment]
adding node 18 was so tempestive that in the meantime our codebase is completely different
